examples/keras_io/tensorflow/nlp/neural_machine_translation_with_transformer.py

```diff
@@ -0,0 +1,477 @@
+import tensorflow.data as tf_data
+
+def format_dataset(eng, spa):
+    eng = eng_vectorization(eng)
+    spa = spa_vectorization(spa)
+    return (
+        {
+            "encoder_inputs": eng,
+            "decoder_inputs": spa[:, :-1],
+        },
+        spa[:, 1:],
+    )
+
+
+def make_dataset(pairs):
+    eng_texts, spa_texts = zip(*pairs)
+    eng_texts = list(eng_texts)
+    spa_texts = list(spa_texts)
+    dataset = tf_data.Dataset.from_tensor_slices((eng_texts, spa_texts))
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(format_dataset)
+    return dataset.shuffle(2048).prefetch(16).cache()
+
+
+train_ds = make_dataset(train_pairs)
+val_ds = make_dataset(val_pairs)
+
+"""
+Let's take a quick look at the sequence shapes
+(we have batches of 64 pairs, and all sequences are 20 steps long):
+"""
+
+for inputs, targets in train_ds.take(1):
+    print(f'inputs["encoder_inputs"].shape: {inputs["encoder_inputs"].shape}')
+    print(f'inputs["decoder_inputs"].shape: {inputs["decoder_inputs"].shape}')
+    print(f"targets.shape: {targets.shape}")
+
+"""
+## Building the model
+
+Our sequence-to-sequence Transformer consists of a `TransformerEncoder`
+and a `TransformerDecoder` chained together. To make the model aware of word order,
+we also use a `PositionalEmbedding` layer.
+
+The source sequence will be pass to the `TransformerEncoder`,
+which will produce a new representation of it.
+This new representation will then be passed
+to the `TransformerDecoder`, together with the target sequence so far (target words 0 to N).
+The `TransformerDecoder` will then seek to predict the next words in the target sequence (N+1 and beyond).
+
+A key detail that makes this possible is causal masking
+(see method `get_causal_attention_mask()` on the `TransformerDecoder`).
+The `TransformerDecoder` sees the entire sequences at once, and thus we must make
+sure that it only uses information from target tokens 0 to N when predicting token N+1
+(otherwise, it could use information from the future, which would
+result in a model that cannot be used at inference time).
+"""
+import keras_core.ops as ops
+
+class TransformerEncoder(layers.Layer):
+    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.dense_dim = dense_dim
+        self.num_heads = num_heads
+        self.attention = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = keras.Sequential(
+            [
+                layers.Dense(dense_dim, activation="relu"),
+                layers.Dense(embed_dim),
+            ]
+        )
+        self.layernorm_1 = layers.LayerNormalization()
+        self.layernorm_2 = layers.LayerNormalization()
+        self.supports_masking = True
+
+    def call(self, inputs, mask=None):
+        if mask is not None:
+            padding_mask = ops.cast(mask[:, np.newaxis, :], dtype="int32")
```

Comment:
`np.newaxis` is just `None` -- better to use `None` explicitly

Reply:
done